Merge

christianbean opened the pull request:
- [x] a minimal gridded perm function

- [x] a new is empty function

enadeau commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -0,0 +1,123 @@
+                         self, gp: GriddedPerm, max_cell_counts: List[Counter],
+                         cells: List[Cell]) -> Iterator[Cell]:
+        """Yield cells that a gridded permutation could be extended by."""
+        currcellcounts = self.cell_counter(gp)
```
For example if you used a `Counter` instead of `defaultdict` here you could just write `currcellcounts = Counter(gp.pos)`

christianbean replied:
Much nicer, I'm sold.

enadeau commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -0,0 +1,123 @@
+                         cells: List[Cell]) -> Iterator[Cell]:
+        """Yield cells that a gridded permutation could be extended by."""
+        currcellcounts = self.cell_counter(gp)
+        if any(all(count >= max_cell_count[cell]
```
Can you explain me why this `all` is not an `any`?

christianbean replied:
A max_cell_count in max_cell_count keeps track of the theoretical bound on the number of points required in each cell in order to create a gridded perm from each requirement list (e.g. if we have R = [[a,b], [c,d]], then there will be a max count for the choices (a, b), (a, c), (b, c), and (b, d).). Therefore this checks that if I am to add another cell, I will definitely be going over the theoretical bound.

enadeau replied:
I get it

enadeau commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -0,0 +1,123 @@
+                for gp in startgp.insert_point(cell):
+                    if gp not in self.seen:
+                        self.seen.add(gp)
+                        if (self.satisfies_obstructions(gp) and
```
You probably have the answer to this. If a permutation have been yielded then it has been seen. Isn't the yielded check redundant with the seen test?

christianbean replied:
Potentially a yielded gridded permutation could be contained in the new gridded perm. As we are not building left to right, but jumping around cells, I can't guarantee there is a unique path to each gridded permutation we come to.

enadeau replied:
Ah I did not realize the **sub** part

christianbean commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -0,0 +1,123 @@
+
+    def minimal_gridded_perms(self) -> Iterator[GriddedPerm]:
+        """Yield all minimal gridded perms on the tiling."""
+        if not self.requirements:
```
Good point

christianbean commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -0,0 +1,123 @@
+        self.queue = []
+        heapify(self.queue)
+
+    def pop(self) -> Info:
```
This is only because I changed the queue system. I agree I can simplify this.

enadeau commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -0,0 +1,180 @@
+
+    @staticmethod
+    def cell_counter(*gps: GriddedPerm) -> Counter:
+        """Returns a counter of the sum of number of cells used by the gridded
```
The docstring is outdated

enadeau commented on tilings/tiling.py:
```diff
@@ -902,7 +904,8 @@ def is_empty(self) -> bool:
             return True
```
I think those two edge cases are now covered by MinimalGriddedPerms

christianbean replied:
I agree, but it prevents creating an MGP instance.

enadeau replied:
Good, not very important anyway

enadeau commented on tilings/tiling.py:
```diff
@@ -933,28 +936,28 @@ def gridded_perms(self,
+                             mgps.minimal_gridded_perms())
+        return self.__class__(self.obstructions, (requirements,))
+
+    def minimal_gridded_perms(self, yield_non_minimal=False):
```
Having this flag is kind of weird. I would remove it and change `is_empty` to
```python
MGP = MinimalGriddedPerms(self)
return any(True for _ in MGP.minimal_gridded_perms(yield_non_minimal))
```

christianbean replied:
Agree, fixed.

enadeau commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -0,0 +1,447 @@
+                    nextgp = gp.insert_specific_point(cell, idx, val)
+                    # The work_packets_done is used to ensure the same
+                    # task is not processed twice.
+                    key = (nextgp, gps, next_cell)
```
I would change the line below to. It would be a bit less nested.
```python 
                    if key in self.work_packets_done:
                        continue
                    self.work_packets_done.add(key)
```

christianbean replied:
Agree, fixed. Also, observed that the attribute was local to this function, so removed it from the class.

enadeau commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -0,0 +1,447 @@
+                        self.work_packets_done.add(key)
+                        # Only work with gridded permutations avoiding the
+                        # obstructions, and those which we haven't yielded a
+                        # subgridded permutation.
```
Also here I would try to use an early `continue`

christianbean replied:
Fixed.

christianbean commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -0,0 +1,447 @@
+        return res
+
+    def get_relevant_obstructions(self, gp: GriddedPerm) -> GPTuple:
+        """Get the requirements that involve only cells in the gp."""
```
Fixed

christianbean commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -0,0 +1,447 @@
+        return res
+
+    def satisfies_obstructions(self, gp: GriddedPerm,
+                               must_contain: Cell = None) -> bool:
```
Fixed

christianbean commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -0,0 +1,447 @@
+    @staticmethod
+    def cell_counter(*gps: GriddedPerm) -> Dict[Cell, int]:
+        """Returns a counter of the sum of number of cells used by the gridded
+        permutations. If 'counter' is given, will update this Counter."""
```
Fixed

christianbean commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -0,0 +1,447 @@
+            # are not involved with any other requirements, because we may be
+            # able to apply a better upper bound to the number of points
+            # required in the cell
+            for cell in set(chain.from_iterable(gp.pos for gp in gps)):
```
Good spot! Fixed.

christianbean commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -0,0 +1,447 @@
+            self.max_cell_counts[gps] = res
+        return res
+
+    _better_bounds = None
```
The logic behind this is that this approach can prevent a long load up time at the start of each run, and it will then only be loaded if necessary.

enadeau replied:
Right, I underestimated the loading time. I thought it would be in nearly instantaneous to import but it actually takes ~1.5s with pypy

jaypantone replied:
If running a whole universe, will it only need to load the first time, or every time?

enadeau replied:
Only the first time. It's cached in the class attribute `MinimalGriddedPerms._better_bounds`

jaypantone replied:
Okay thanks, so different instances of MinimalGriddedPerms will all be accessing the same copy of the class attribute. Makes sense!

christianbean commented:
> Wow, I'm very impressed by the amount of caching going on in this. So many dictionaries.
In fact the keys are almost always the same ( 5 of the 9 are FrozenSet[Cell], and 2 more are Tuple[Cell, FrozenSet[cell]] ). 
In fact, the following
```python
        self.localised_patts = dict()  # Dict[GPTuple, GPTuple]
        self.max_cell_counts = dict()  # Dict[GPTuple, Dict[Cell, int]]
        self.upward_closures = dict()  # Dict[GPTuple, GPTuple]
        self.known_patts = defaultdict(set)  # DefaultDict[GriddedPerm, Set[GriddedPerm]]
```
are not tiling specific, so could be stored on the class altogether, allowing there corresponding methods to be staticmethods (and in turn also the methods `initial_gp` which relies on `get_upward_closure`).

> Two general comments:
> 
> * I think we should hide the method that are no intended to be used outside the class by prefixing them with a `_`

Agree. I applied this to the methods `get_cells_to_try`, `prepare_queue`. I also moved the function `yielded_subgridded_perm` to be a sub function of `minimal_gridded_perms` as in fact, then the attributes `yielded` and `initial_gps_to_auto_yield` become local to the function, and also this ensures that `minimal_gridded_perms` can be called twice without odd behaviour. 

> * I'm a bit afraid of the `minimal_gridded_perms`. It's very big and deeply nested as pointed out by the linter. I must say that I don't really understand it. I've tried to pin point some easy fix to make it less complex to read but I think it would be very good if we were able to break it into smaller functions.

I removed one layer of complexity by moving the point insertion to a separate function, and use more continues as the specific comments suggested. Do you think it is still a bit monstrous? I think the amount of comments make it look like more.

enadeau commented on tilings/algorithms/minimal_gridded_perms.py:
```diff
@@ -301,13 +285,13 @@ def get_upward_closure(self, gps: GPTuple) -> GPTuple:
+                          gps: GPTuple,
+                          last_cell: Cell,
+                          try_localised: bool
+                         ) -> Iterator[Cell]:
```
The output type is incorrect as it returns a `Tuple[Cell, bool]`

christianbean replied:
Fixed.
